domains/pallets/messenger/src/lib.rs

```diff
@@ -1510,14 +1511,54 @@ where
         Some(T::create_unsigned(call.into()))
     }
 
-    /// Returns true if the outbox message has not received the response yet.
-    pub fn should_relay_outbox_message(dst_chain_id: ChainId, msg_id: MessageId) -> bool {
-        Outbox::<T>::contains_key((dst_chain_id, msg_id.0, msg_id.1))
+    /// Returns the first outbox message nonce that should be relayed to the dst_chain.
+    pub fn should_relay_outbox_messages(
+        dst_chain_id: ChainId,
+        channel_id: ChannelId,
+        from_nonce: Nonce,
+    ) -> Option<Nonce> {
+        Self::should_relay_message(
+            dst_chain_id,
+            channel_id,
+            from_nonce,
+            Outbox::<T>::contains_key,
+        )
     }
 
-    /// Returns true if the inbox message response has not received acknowledgement yet.
-    pub fn should_relay_inbox_message_response(dst_chain_id: ChainId, msg_id: MessageId) -> bool {
-        InboxResponses::<T>::contains_key((dst_chain_id, msg_id.0, msg_id.1))
+    /// Returns the first inbox response message nonce that should be relayed to the dst_chain.
+    pub fn should_relay_inbox_message_responses(
+        dst_chain_id: ChainId,
+        channel_id: ChannelId,
+        from_nonce: Nonce,
+    ) -> Option<Nonce> {
+        Self::should_relay_message(
+            dst_chain_id,
+            channel_id,
+            from_nonce,
+            InboxResponses::<T>::contains_key,
```

Comment:
It will be much easier if there is a `lowest_pending_msg_nonce` field in the state, tracking the oldest pending msg's nonce, so we only need one db read instead of potentially 256.

Reply:
I did think on this earlier.
Theoretically true but this also adds extra storage cost for each XDM that is included if we have such a storage either after every extrinsic or during the finalization of block. 

While this may potentially loop for 256 messages, practically this will not have such many loops since relayer only starts after major sync. 

I do not see a big boost coming from this at the cost of extra storage items and multiple writes per block.

I would prefer not adding such, at least in this iteration